File: src/kotlin/sequential/graph/dijkstra/Dijkstra.kt
package sequential.graph.dijkstra


// Weighted graph map representation where `key` = source vertex, `value` = destination vertices with edge weights.
private typealias Graph<T> = Map<T, Edges<T>>
// Weighted edges to destination vertices.
private typealias Edges<T> = Map<T, Weight>
// Edge weight.
private typealias Weight = Double


// Infinity as a default vertex cost value.
private val INFINITY = Double.POSITIVE_INFINITY


/**
 * Finds shortest path to the [target] node using Dijkstra algorithm.
 */
fun <T: Any> Graph<T>.dijkstra(root: T, target: T): Collection<T> {
    val graph = this
    require(root in graph) { "Invalid root vertex." }

    val costs = graph[root]!!.toMutableMap()
    val explored = mutableSetOf<T>()
    val parents = graph[root]!!.mapValues { root }.toMutableMap()

    var node = costs.findMinCostNode(explored)
    while (node != null) {
        explored += node
        val nodeCost = costs[node]!!

        val successors = graph[node].orEmpty()
        for ((succ, edgeWeight) in successors) {
            if (nodeCost + edgeWeight < costs[succ] ?: INFINITY) {
                costs[succ] = nodeCost + edgeWeight
                parents[succ] = node
            }
        }
        node = costs.findMinCostNode(explored)
    }

    return buildPathToNode(target, parents)
}

private fun <T> Map<T, Double>.findMinCostNode(explored: Collection<T>): T? = 
    // turning costs to sequence for faster further operations.
    asSequence()
        // ignoring already explored nodes.
        .filter { (node, _) -> node !in explored } 
        // finding the min cost node.
        .minBy { (_, cost) -> cost} 
        // returning node object or `null` if the node was not found.
        ?.key 

/**
 * Build path from root to target node.
 */
private fun <T: Any> buildPathToNode(target: T, parents: Map<T, T>): List<T> {
    require(target in parents) { "Invalid target vertex."}

    return generateSequence(target) { node -> parents[node] }
        .toList()
        .asReversed()
}


fun main() {
    val graph = mapOf(
        "Start" to mapOf("A" to 5.0),
        "A" to mapOf("B" to 7.0, "C" to 4.0),
        "B" to mapOf("Finish" to 4.0),
        "C" to mapOf("Finish" to 3.0)
    )

    println(
        graph.dijkstra("Start", "Finish")
    )
}

File: src/swift/sequential/graph/dijkstra/Dijkstra.swift
/// Weighted graph representation using dictionary.
typealias Graph<T: Hashable> = [T: [T: Double]]


/// Infinity as a default node cost value.
let infinity = Double.infinity


/**
 Finds fastest path between specified nodes in the given graph.

 - Parameters:
    - graph: Given graph.
    - root: Beginning of the path.
    - target: End of the path.

 - Returns: Array representing path to the `target` node starting from `root` or `nil` if invalid arguments provided.
 */
func dijkstra<T>(in graph: Graph<T>, withRoot root: T, findPathTo target: T) -> [T]? {
    guard graph[root] != nil else { return nil }

    var costs = graph[root]!
    var explored = Set<T>()
    var parents = graph[root]!.mapValues { _ in root }
    
    while let node = findMinCostNode(in: costs, except: explored) {
        explored.insert(node)
        let nodeCost = costs[node]!
        
        guard let successors = graph[node] else { continue }

        for (succ, edgeWeight) in successors {
            if nodeCost + edgeWeight < costs[succ] ?? infinity {
                costs[succ] = nodeCost + edgeWeight
                parents[succ] = node
            }
        }
    }
    
    return buildPathTo(target, with: parents)
}

/**
 Finds the cheapest node at the given moment.
 
 - Parameters:
    - costs: Dictionary which contains costs of all nodes at the given moment.
    - explored: Ignored nodes which are already explored.
 
 - Returns: Min cost node at the given moment or `nil` if all nodes are already explored.
 */
private func findMinCostNode<T>(in costs: [T: Double], except explored: Set<T>) -> T? {
    return costs
        // ignoring already explored nodes.
        .filter { node, cost in !explored.contains(node) }
        // finding the min cost node. `value` corresponds to node cost.
        .min { cost1, cost2 in cost1.value < cost2.value}?
        // returning node object or `nil` if the min node was not found.
        .key
}

/**
 Builds a path to the given node starting from root.
 
 - Parameters:
    - target: Given node which represends the end of the path.
    - parents: Dictionary which contains parents of all graph nodes.
 
 - Returns: Array representing path to the `target` node starting from `root` or `nil` if invalid arguments provided.
 */
private func buildPathTo<T>(_ target: T, with parents: [T: T]) -> [T]? {
    guard parents[target] != nil else { return nil }
    
    var path = [T]()
    var node: T? = target
    while node != nil {
        path.append(node!)
        node = parents[node!]
    }
    return path.reversed()
}


func mainDijkstra() {
    let graph = [
        "Start": ["A": 5.0],
        "A": ["B": 7.0, "C": 4.0],
        "B": ["Finish": 4.0],
        "C": ["Finish": 3.0]
    ]

    print(
        dijkstra(in: graph, withRoot: "Start", findPathTo: "Finish") ?? "Path not found!"
    )
}